### ISSUE TITLE
What is the reason for using `fetchDependencies` over git submodules?

### ISSUE THREAD

**Issue by ioquatix** [KhronosGroup/MoltenVK#499]
It seems like using git submodules would be more portable and straight forward than a custom script.

**Comment by billhollings** [comment]
MoltenVK originally used git submodules. They were changed to a script-based system for flexibility:

1. Submodules cannot invoke build behaviour...so some form of `buildDepenencies` script would still be needed.

2. The `fetchDependencies` script allows an external build environment (like the Vulkan SDK) to override the version of dependencies that are common to MoltenVK and the external build environment.

3. The script allows for the potential of partial retrievals, builds, etc.

4. The script and file-based revision info provides convenient editing and traceability for dependency versions.

5. The script can provide extended logging and validation of dependency processing.

**Comment by ioquatix** [comment]
Thanks. That all makes sense.

What I did in the end was compile everything and only use the build artefacts into my own source tree.

I wanted to pull in the entire dependency so I could compile it with different flags, e.g. arch, optimisation, etc.

The way this code is set up (e.g. symlinks) makes it very hard to do an out-of-source build.

Including fetching dependencies as part of the build process makes it much more tricky.

I really love what you've achieved here. It's just very hard to integrate into other build systems, unless you basically just do it manually and copy the static library/headers.

**Comment by billhollings** [comment]
Yeah. As an Apple-only product, we use _Xcode_ to stay current with the evolution of the Metal SDK.

For the libraries built in `fetchDependencies`, you can modify the build settings in the `ExternalDependencies.xcodeproj` _Xcode_ project...and then trigger the `fetchDependencies` script in your build environment.

The building of the dependency libraries was separated from the main MoltenVK product to speed up the build time for MoltenVK, whose code might change at different rates that the code in the dependency libraries...especially for contributors to the repo.
